toxcore/crypto_core.cpp

```diff
@@ -0,0 +1,245 @@
+{
+    /* FIXME use increment_nonce_number(nonce, 1) or sodium_increment (change to little endian)
+     * NOTE don't use breaks inside this loop
+     * In particular, make sure, as far as possible,
+     * that loop bounds and their potential underflow or overflow
+     * are independent of user-controlled input (you may have heard of the Heartbleed bug).
+     */
+    uint32_t i = crypto_box_NONCEBYTES;
+    uint_fast16_t carry = 1U;
+    for (; i != 0; --i) {
+        carry += (uint_fast16_t) nonce[i - 1];
+        nonce[i - 1] = (uint8_t) carry;
+        carry >>= 8;
+    }
+}
+
+void increment_nonce_number(uint8_t* nonce, uint32_t host_order_num)
+{
+    /* NOTE don't use breaks inside this loop
+     * In particular, make sure, as far as possible,
+     * that loop bounds and their potential underflow or overflow
+     * are independent of user-controlled input (you may have heard of the Heartbleed bug).
+     */
+    const uint32_t big_endian_num = htonl(host_order_num);
+    const uint8_t* const num_vec = reinterpret_cast<const uint8_t*>( &big_endian_num );
+    uint8_t num_as_nonce[crypto_box_NONCEBYTES] = {0};
+    num_as_nonce[crypto_box_NONCEBYTES - 4] = num_vec[0];
+    num_as_nonce[crypto_box_NONCEBYTES - 3] = num_vec[1];
+    num_as_nonce[crypto_box_NONCEBYTES - 2] = num_vec[2];
+    num_as_nonce[crypto_box_NONCEBYTES - 1] = num_vec[3];
+
+    uint32_t i = crypto_box_NONCEBYTES;
+    uint_fast16_t carry = 0U;
+    for (; i != 0; --i) {
+        carry += (uint_fast16_t) nonce[i] + (uint_fast16_t) num_as_nonce[i];
+        nonce[i] = (unsigned char) carry;
+        carry >>= 8;
+    }
+}
+
+void random_nonce(uint8_t* nonce)
+{
+    randombytes_buf(nonce, crypto_box_NONCEBYTES);
+}
+
+void new_nonce(uint8_t* nonce)
+{
+    random_nonce(nonce);
+}
+
+void new_symmetric_key(uint8_t* key)
+{
+    randombytes(key, crypto_box_BEFORENMBYTES);
+}
+
+template<class PointerT>
+struct Packet_Pointers
+{
+    PointerT const type;
+    PointerT const recv_public_key;
+    PointerT const send_public_key;
+    PointerT const nonce;
+    PointerT const cyphertext;
+
+    explicit Packet_Pointers(PointerT const packet):
+        type( packet + 0 ),
+        recv_public_key( packet + 1 ),
+        send_public_key( packet + 1 + crypto_box_PUBLICKEYBYTES ),
+        nonce( packet + 1 + crypto_box_PUBLICKEYBYTES * 2 ),
+        cyphertext( packet + 1 + crypto_box_PUBLICKEYBYTES * 2 + crypto_box_NONCEBYTES)
+    {}
+private:
+    Packet_Pointers(); // = delete
+    Packet_Pointers(const Packet_Pointers&); // = delete
+    Packet_Pointers& operator= (const Packet_Pointers&); // = delete
+};
+
+int create_request(const uint8_t* send_public_key, const uint8_t* send_secret_key, uint8_t* packet,
+                   const uint8_t* recv_public_key, const uint8_t* data, uint32_t length, uint8_t request_id)
+{
+    if (MAX_CRYPTO_REQUEST_SIZE < length + 1 + crypto_box_PUBLICKEYBYTES * 2 + crypto_box_NONCEBYTES + 1 + crypto_box_MACBYTES
```

Comment:
It would be better to simplify the expression by extracting math into constant and use it in condition.